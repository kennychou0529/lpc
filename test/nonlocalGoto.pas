
// non-local goto tests

program NonlocalGoto;

label 1, 2;

var p1, p2 : integer;

procedure bar(x : integer); forward;

function foo(x : integer) : integer;
begin
    case x of
    1: bar(x);
    3: goto 1; // non-local P_L1
    2: goto 2; // non-local P_L2
    end;
    writeln('UNEXPECTED');
    foo := -1;
end; 

procedure bar;

    label 1;

    procedure moo;
    begin
        goto 1; // BAR_L1
    end;

begin
    moo;
    writeln('UNEXPECTED');
1: // BAR_L1
    p2 := foo(3);
end;

begin
    p1 := 0;
    p2 := 0;  

1: // P_L1
    writeln('P_L1');
    writeln('p1 = ', p1);
    writeln('p2 = ', p2);

    p1 := p1 + 1;

    p2 := foo(p1);
    writeln('UNEXPECTED');
    goto 1; // local goto P_L1

2: // P_L2
    writeln('P_L2');
    writeln('p1 = ', p1);
    writeln('p2 = ', p2);
end.

