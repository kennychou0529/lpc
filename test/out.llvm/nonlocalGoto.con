
compiling pass 1 (LPC)
-----------------------------------------------------
  warning on line 16: non-local goto
  warning on line 17: non-local goto
  warning on line 29: non-local goto

0 error(s), 3 warning(s)

P1 errcode = 0

compiling pass 2 (target = llvm)
-----------------------------------------------------

P2 errcode = 1
