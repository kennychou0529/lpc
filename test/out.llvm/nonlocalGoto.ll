
; line 0
;================================================================================
; scope: global scope (level : 0)

; types
%T_pointer_1 = type i8*
%T_text = type <file>

; line 4
;================================================================================
; scope: program scope (level : 1)

; types

; program variables
@_input = dso_local global %T_text zeroinitializer
@_output = dso_local global %T_text zeroinitializer
@p1 = dso_local global i32 zeroinitializer
@p2 = dso_local global i32 zeroinitializer

; line 12
;================================================================================
; scope: foo (level : 2)

; types

; line 23
;================================================================================
; scope: bar (level : 2)

; types

; line 27
;================================================================================
; scope: bar_moo (level : 3)

; types
