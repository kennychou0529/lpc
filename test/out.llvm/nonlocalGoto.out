P_L1
p1 = 0
p2 = 0
P_L1
p1 = 1
p2 = 0
P_L2
p1 = 2
p2 = 0
